torch/_dynamo/variables/torch.py

```diff
@@ -773,13 +773,14 @@ def speculate_subgraph(
             output = f.call_function(tx, args, {})
             # Register output to graph
             # Modeled off of compile_and_call_fx_graph
-            # TODO: support non single Tensor output
+            # TODO: support pytree output
             # We check always_restore because we dont use the output or side effects of always_restore code,
             # like bwd.
-            if not isinstance(output, TensorVariable) and not always_restore:
-                unimplemented(
-                    "HigherOrderOperator with body with non single Tensor output"
-                )
+            if (
+                not isinstance(output, (TensorVariable, ListVariable, TupleVariable))
+                and not always_restore
+            ):
+                unimplemented("HigherOrderOperator with body with pytree output")
```

Comment:
Do we need a check that if we are dealing with a TupleVariable (or a ListVariable), that it contains only TensorVariable?

(Also, could we add a test for nested tuple? I would expect that to graph break)

Reply:
yeah, you're right. For output, we do want to make sure they only consist of TensorVariable. Will add check and corresponding test for it. 

Nested tuple case is a bit tricky, there's no graph breaks because ListVariable/TupleVariable.as_proxy() will do as_proxy() recursively for elements in the list/tuple. The end result is like this:
```python
1. class GraphModule(torch.nn.Module):
2.     def forward(self, L_x_ : torch.Tensor):
3.         l_x_ = L_x_
4.         
5.         # File: /scratch/yidi/work/pytorch/test/dynamo/test_higher_order_ops.py:423, code: return wrap(lambda x: [torch.sin(x), torch.cos(x), [-x, -x]], x)
6.         wrap_body_0 = self.wrap_body_0
7.         wrap = torch._higher_order_ops.wrap.wrap(wrap_body_0, l_x_);  wrap_body_0 = l_x_ = None
8.         getitem = wrap[0]
9.         getitem_1 = wrap[1]
10.        getitem_2 = wrap[2];  wrap = None
11.        getitem_3 = getitem_2[0]
12.        getitem_4 = getitem_2[1];  getitem_2 = None
13.        return (getitem, getitem_1, getitem_3, getitem_4)
14.         
15.     class GraphModule(torch.nn.Module):
16.         def forward(self, l_x_):
17.             # File: /scratch/yidi/work/pytorch/test/dynamo/test_higher_order_ops.py:423, code: return wrap(lambda x: [torch.sin(x), torch.cos(x), [-x, -x]], x)
18.             sin = torch.sin(l_x_)
19.             cos = torch.cos(l_x_)
20.             neg = -l_x_
21.             neg_1 = -l_x_;  l_x_ = None
22.             return [sin, cos, [neg, neg_1]]
```

The problem with this graph is that its output is not flattened. Adding a test doesn't seem to make a lot of sense in this case..

Reply:
> (Also, could we add a test for nested tuple? I would expect that to graph break)

What I meant by this was add a test checking if `wrap`'s body_function returns a nested tuple then Dynamo correctly graph breaks on it instead of trying to process it. This should be the behavior until we figure out the pytree business

Reply:
I got your point. Should we turn non-tensor check into RunTimeError? In that case, the nested list case would be treated as RuntimeError.